### ISSUE TITLE
CloseDB error apparently triggered by peer disconnect.

### ISSUE THREAD

**Issue by karknu** [input-output-hk/ouroboros-network#1350]
A cardano-node disconnecting (“resource vanished (Broken pipe)” line) triggered a CloseDBE rror which brought down the node. byron-proxy  was started with -N2. A mux bearer can be cancelled at any time, and it shouldn’t cascade into an error that brings down the node (or proxy).

Log output:
```
[cardano_byron_proxy.diffusion.outboundqueue.self:Debug:80] [2019-12-12 14:53:16.07 UTC] inFlight = fromList []
[cardano_byron_proxy.diffusion.outboundqueue.self:Debug:80] [2019-12-12 14:53:16.07 UTC] message MsgRequestBlocks (fromList [NodeId 54.64.0.53:3000:0]) enqueued to [NodeId 54.64.0.53:3000:0]
[cardano_byron_proxy.diffusion.outboundqueue.self:Debug:85213] [2019-12-12 14:53:16.07 UTC] sending MsgRequestBlocks (fromList [NodeId 54.64.0.53:3000:0]) to NodeId 54.64.0.53:3000:0
[cardano_byron_proxy.shelley.server:Error:62] [2019-12-12 14:53:16.53 UTC] IP 192.168.1.100:46729 ErrorPolicyUnhandledApplicationException (MuxError {errorType = MuxIOException writev: resource vanished (Broken pipe), errorMsg = "(sendAll errored)", errorStack = []})
[cardano_byron_proxy.diffusion:Debug:85216] [2019-12-12 14:53:16.95 UTC] streamBlocks: Requesting stream of blocks from [b75400dfa52ec1fe, 2b934653e715dd87, e14f2c0e3e0e6b7e, 2a8aba44f7d7f3bf, 0a7a4395d81fdf14, 0164991d90a36b19, 940ae3fc58af8e76, 4637394e1cca4a5c, c9af9173247c8085, a7de1d35da47fef4, db0343773873edf3, e84bcd3b3a832089, 5a392fe4ab13dc38, 4aa86a1b1488e574, ff6ebaddcc93c6bc, caa9b595d6b60824, dbe9124986bcfd5d, 665e6284ea31889e] to 5bdb43a2
[cardano_byron_proxy.diffusion:Debug:85219] [2019-12-12 14:53:17.84 UTC] Streaming done client-side for node"54.64.0.53:3000:0"
[cardano_byron_proxy.ChainDB:Info:85220] [2019-12-12 14:53:17.84 UTC] blockNum = 3495908
[cardano_byron_proxy.diffusion.outboundqueue.self:Debug:85213] [2019-12-12 14:53:17.84 UTC] sent MsgRequestBlocks (fromList [NodeId 54.64.0.53:3000:0]) to NodeId 54.64.0.53:3000:0
[cardano_byron_proxy.index:Info:50] [2019-12-12 14:53:17.84 UTC] Rollforward (Block {blockPointSlot = SlotNo {unSlotNo = 3497965}, blockPointHash = ByronHash {unByronHash = AbstractHash 5bdb43a282a724a5c0e78a22ef0df68d45cc55c208c6cb64142e608a1b09393f}})
[cardano_byron_proxy.diffusion:Error:59] [2019-12-12 14:53:17.84 UTC] stopping with exception AsyncCancelled
cardano-byron-proxy: ExceptionInLinkedThread "ThreadId 50" ClosedDBError
```

**Comment by mrBliss** [comment]
The background thread spawned here: https://github.com/input-output-hk/cardano-byron-proxy/blob/7a7d93868cd0c3c78079d29ae56d2684e6cfbfb2/src/exec/DB.hs#L67 should be spawned using `withAsync`, but then it can no longer access the `ResourceRegistry` from the new thread, as it is not registered in the registry.

So what we should is provide a `withThread` function in `ResourceRegistry`, which is basically `withAsync`, but with the thread registered in the registry so it can allocate resources in the registry.

**Comment by mrBliss** [comment]
Actually, @edsko pointed out that this ticket is not about `CloseDBError` (I just glanced over the trace).

The problem is the following line:
```
[cardano_byron_proxy.shelley.server:Error:62] [2019-12-12 14:53:16.53 UTC] IP 192.168.1.100:46729 ErrorPolicyUnhandledApplicationException (MuxError {errorType = MuxIOException writev: resource vanished (Broken pipe), errorMsg = "(sendAll errored)", errorStack = []})
```
I interpreted this as: the byron-proxy's error policy is incomplete and should handle this exception too.

The `CloseDBError` is a totally different problem, but it is triggered whenever the proxy is shut down.

**Comment by mrBliss** [comment]
@karknu Can you create ticket in https://github.com/input-output-hk/cardano-byron-proxy/ for the incomplete error policy?

The `CloseDBError` side of this issue, we'll handle, but it is low priority, because the node is still shut down. If we fix that side of the issue, it will just shut down a little more cleanly.

**Comment by karknu** [comment]
I don't think this is a byron-proxy only problem.
From the slack discussion when the ticket was filed:
>Karl Knutsson  27 days ago
Yes, the byron has proper error policies. I’ve also seen this problem for cardano-node twice when running on Arm but didn’t report it due to the heap corruption problem on that platform.

See https://input-output-rnd.slack.com/archives/CFKLUH4R0/p1576163214173500

**Comment by karknu** [comment]
@mrBliss Could it be that #1452 fixes this too? A disconnecting peer will trigger an asynchronous interrupt.

**Comment by mrBliss** [comment]
> @mrBliss Could it be that #1452 fixes this too? A disconnecting peer will trigger an asynchronous interrupt.

I think that #1448 (PR #1449) is more likely to fix this. 

Maybe you can verify it (by cherry-picking that commit)?

**Comment by karknu** [comment]
I'll see if I can find a way to reliably trigger it. Then cherry-picking and testing would work.

**Comment by mrBliss** [comment]
@karknu Did you find a way to trigger this? I'm trying to reproduce it with `cardano-node-1.1.0` and `cardano-byron-proxy`: I'm running the proxy and letting the node sync from it but killing the node using the `timeout` command in a loop. I see the following message in the proxy log each time the node is killed:
```
cardano_byron_proxy.shelley.server:Error:58] [2020-02-07 10:48:53.73 UTC] IP 127.0.0.1:41583 ErrorPolicyUnhandledApplicationException (MuxError {errorType = MuxIOException writev: resource vanished (Connection reset by peer), errorMsg = "(sendAll errored)", errorStack = []})
```
but the proxy keeps on running. Maybe it depends on the exact timing?

I still don't know how this could happen, regardless of #1449 :confused:

**Comment by karknu** [comment]
@mrBliss The last time I tried to trigger this I ran into random livelocks and segfaults. This turned out to be https://gitlab.haskell.org/ghc/ghc/merge_requests/734 .

I think that in order to trigger this you will need to run with `-N` > 1 and the serving proxy or cardano-node should be in the process of initial chainsyncing while at the same having a restarting client chainsyncing from it. 

I will setup a non-Arm environment next week and see if I can trigger it from that.

**Comment by karknu** [comment]
I have a setup with two nodes, one fully synced and one starting fresh and a byron-proxy performing an initial chainsync. The fully synced node will connect and reconnect every second and the node starting fresh is killed an restarted every minute.
I've managed to trigger the fault twice within 2h of starting the proxy on version 1.0.2 but so far 1.0.4 is running fine.

**Comment by karknu** [comment]
@mrBliss this error was fixed by #1389.
